BEIMA.Client/src/pages/Devices/AddDevicePage.js

```diff
@@ -62,83 +72,151 @@ const AddDevicePage = () => {
+          newErrors[formName] = `${formName} value is invalid. Must be a decimal between -${coordMax} and ${coordMax}.`;
+        }
+      }
+      
+      //construct the current JSON element to be added to the insert
       if(fieldNames.includes(formName)){
-        let formJSON =  {[formName] : formFields[i].value};
+        let formJSON = '';
+        let jsonKey = convertToDbFriendly(formName);
         
-        //lat lon validation
-        if (formName === 'Latitude' || formName === 'Longitude') {
-          const coordMax = formName === 'Latitude' ? 90 : 180;
-          if(!(isFinite(formFields[i].value) && Math.abs(formFields[i].value) <= coordMax)) {
-            newErrors[formName] = `${formName} value is invalid. Must be a decimal between -${coordMax} and ${coordMax}.`;
-          }
+        //straight append when it's not part of a nested element
+        //otherwise it attaches it to the nested element
+        if(!(typeof jsonKey == 'object')){
+          formJSON =  {[jsonKey] : deviceFields[formName]};
+          Object.assign(dbJson, formJSON);
+        } else if ('fieldDbId' in jsonKey){
+          dbJson.fields[jsonKey.fieldDbId] = deviceFields[formName];
+        } else if ('location' in jsonKey){
+          dbJson.location[jsonKey.location.type] = deviceFields[formName];
         }
-        
-        Object.assign(fieldValues, formJSON);
       } else if (formName === "Device Image"){
         setDeviceImage(formFields[i].files[0]);
-        formFields[i].value = "";
       } else if (formName === "Additional Documents"){
         setAdditionalDocs(formFields[i].files);
-        formFields[i].value = "";
       }
     }
-
-    setDeviceFields(mandatoryDeviceFields);
-    setFullDeviceJSON(fieldValues);
     
     //won't deploy to azure without these until they're used elsewhere
-    console.log(fullDeviceJSON);
     console.log(deviceImage);
     console.log(deviceAdditionalDocs);
-
+    
+    //display errors when present or attempt insert when valid data is present
     if ( Object.keys(newErrors).length > 0 ) {
       setErrors(newErrors);
     } else {
-      setErrors({});
-      setSelectedDeviceType('Select Device Type');
-      for(let i = 0; i < formFields.length; i++){
-        formFields[i].value = "";
-      }
+      /*TODO Need ability to get a building ID, probably a DD on the page*/
+
+      dbJson.deviceTypeId = selectedDeviceType._id;
       
+      AddDevice(dbJson).then(response => {
+        //reset or show problem
+        if(response.status === HTTP_SUCCESS){
+          setErrors({});
+          setSelectedDeviceType(defaultDeviceTypeObj);
+          setDdStyle(styles.ddSelected);
+          for(let i = 0; i < formFields.length; i++){
+            formFields[i].value = "";
+          }
+        } else {
+          /*TODO push error to display*/
+          alert('API responded with: ' + response.status + ' ' + response.response);
```

Comment:
It is likely that our users won't understand what this means, we should say something like 'Adding the device failed: response details...' instead of 'API responded with'.

Reply:
The signaling can be improved, I don't like the alert function either it's basic and looks unpolished. Since this coming sprint is "longer" because of the break I was thinking of getting a status/error section inserted into the top of the page. I was thinking a shared component that can be added to any page that needs it with some styling.

Even the response details are sparse from the API (status 400, text Bad Request) so the wording/suggested action will need some massaging. 

Some tasks can be added to #157 for the next sprint to polish it, but at the moment I feel it satisfies the goal of the task to get it working with the API calls to add a device.

We'll need a task to update it for the building part too since it was only built last sprint as well.